File: topics/examples/nats-client-example.py
"""
NATS Client示例代码 (Python)

功能：演示NATS客户端的基本用法和最佳实践
参考：[01-05-程序设计模式分析](../01-基础概念与对比分析/01-05-程序设计模式分析.md)
依赖：pip install nats-py
"""

import asyncio
import logging
from nats.aio.client import Client as NATS
from nats.aio.errors import ErrConnectionClosed, ErrTimeout

# 配置日志
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class NATSClientExample:
    def __init__(self, servers=['nats://localhost:4222']):
        self.nc = NATS()
        self.servers = servers
    
    async def connect(self):
        """连接到NATS服务器"""
        try:
            await self.nc.connect(
                servers=self.servers,
                reconnect_time_wait=1,  # 重连等待时间
                max_reconnect_attempts=-1,  # 无限重连
                name='python-example-client',
            )
            logger.info("NATS连接成功")
        except Exception as e:
            logger.error(f"NATS连接失败: {e}")
            raise
    
    async def publish_subscribe(self):
        """发布-订阅模式（参考：01-05-程序设计模式分析）"""
        subject = "events.user.created"
        
        # 订阅消息
        async def message_handler(msg):
            logger.info(f"收到消息: subject={msg.subject}, data={msg.data.decode()}")
            # 处理消息逻辑
            await self.process_message(msg)
        
        sub = await self.nc.subscribe(subject, cb=message_handler)
        logger.info(f"已订阅主题: {subject}")
        
        # 发布消息
        await self.nc.publish(subject, b"user created: 12345")
        logger.info(f"已发布消息到主题: {subject}")
        
        # 等待消息处理
        await asyncio.sleep(1)
        
        # 取消订阅
        await sub.unsubscribe()
    
    async def request_reply(self):
        """请求-响应模式（参考：02-03-程序设计模式场景化应用）"""
        subject = "service.echo"
        
        # 订阅请求
        async def request_handler(msg):
            response = f"Echo: {msg.data.decode()}"
            await msg.respond(response.encode())
            logger.info(f"处理请求: {msg.data.decode()}, 响应: {response}")
        
        await self.nc.subscribe(subject, cb=request_handler)
        logger.info(f"已订阅请求主题: {subject}")
        
        # 发送请求
        try:
            response = await self.nc.request(subject, b"hello", timeout=1)
            logger.info(f"收到响应: {response.data.decode()}")
        except ErrTimeout:
            logger.error("请求超时")
    
    async def queue_group(self):
        """队列组模式（负载均衡）"""
        subject = "tasks.process"
        queue_group = "worker-group"
        
        # 多个Consumer订阅同一Subject，消息只投递给一个Consumer
        async def task_handler(msg):
            logger.info(f"处理任务: {msg.data.decode()}")
            # 处理任务逻辑
            await self.process_task(msg)
        
        await self.nc.subscribe(subject, queue=queue_group, cb=task_handler)
        logger.info(f"已加入队列组: {queue_group}, 主题: {subject}")
        
        # 发布任务
        for i in range(5):
            await self.nc.publish(subject, f"task-{i}".encode())
        
        await asyncio.sleep(1)
    
    async def jetstream_example(self):
        """JetStream持久化示例"""
        js = self.nc.jetstream()
        
        # 创建Stream
        stream_name = "ORDERS"
        await js.add_stream(
            name=stream_name,
            subjects=["orders.>"],
            max_age=3600,  # 1小时
            max_bytes=1024 * 1024 * 100,  # 100MB
        )
        logger.info(f"已创建Stream: {stream_name}")
        
        # 发布消息到Stream
        await js.publish("orders.new", b"order-123")
        logger.info("已发布消息到Stream")
        
        # 创建Consumer并消费消息
        async def msg_handler(msg):
            logger.info(f"收到Stream消息: {msg.data.decode()}")
            await msg.ack()
        
        await js.subscribe("orders.>", cb=msg_handler)
        await asyncio.sleep(1)
    
    async def process_message(self, msg):
        """处理消息"""
        # 业务逻辑处理
        pass
    
    async def process_task(self, msg):
        """处理任务"""
        # 任务处理逻辑
        pass
    
    async def close(self):
        """关闭连接"""
        await self.nc.close()


async def main():
    """主函数"""
    client = NATSClientExample(['nats://localhost:4222'])
    
    try:
        # 连接
        await client.connect()
        
        # 发布-订阅模式
        await client.publish_subscribe()
        
        # 请求-响应模式
        await client.request_reply()
        
        # 队列组模式
        await client.queue_group()
        
        # JetStream示例
        await client.jetstream_example()
        
    except Exception as e:
        logger.error(f"异常: {e}")
    finally:
        await client.close()


if __name__ == '__main__':
    asyncio.run(main())

File: topics/examples/nats-client-example.go
/**
 * NATS Client示例代码
 *
 * 功能：演示NATS客户端的基本用法和最佳实践
 * 参考：[02-03-程序设计模式场景化应用](../02-场景驱动架构设计/02-03-程序设计模式场景化应用.md)
 */

package main

import (
	"fmt"
	"log"
	"time"

	"github.com/nats-io/nats.go"
)

func main() {
	// 1. 连接到NATS服务器（支持多个服务器地址，自动故障转移）
	nc, err := nats.Connect("nats://localhost:4222",
		nats.Name("example-client"),
		nats.ReconnectWait(1*time.Second),
		nats.MaxReconnects(10),
		nats.DisconnectErrHandler(func(nc *nats.Conn, err error) {
			if err != nil {
				log.Printf("连接断开: %v", err)
			}
		}),
		nats.ReconnectHandler(func(nc *nats.Conn) {
			log.Printf("重连成功")
		}),
	)
	if err != nil {
		log.Fatal(err)
	}
	defer nc.Close()

	// 2. 发布-订阅模式（参考：01-05-程序设计模式分析）
	subject := "events.user.created"

	// 订阅消息（事件驱动模式）
	sub, err := nc.Subscribe(subject, func(msg *nats.Msg) {
		fmt.Printf("收到消息: subject=%s, data=%s\n", msg.Subject, string(msg.Data))
		// 处理消息逻辑
		processEvent(msg)
	})
	if err != nil {
		log.Fatal(err)
	}
	defer sub.Unsubscribe()

	// 发布消息
	nc.Publish(subject, []byte("user created: 12345"))

	// 3. 请求-响应模式（参考：02-03-程序设计模式场景化应用）
	// 订阅请求
	nc.Subscribe("service.echo", func(msg *nats.Msg) {
		response := fmt.Sprintf("Echo: %s", string(msg.Data))
		msg.Respond([]byte(response))
	})

	// 发送请求
	response, err := nc.Request("service.echo", []byte("hello"), 1*time.Second)
	if err != nil {
		log.Printf("请求失败: %v", err)
	} else {
		fmt.Printf("收到响应: %s\n", string(response.Data))
	}

	// 4. 队列组模式（负载均衡）
	// 多个Subscriber订阅同一Subject，消息只投递给一个Subscriber
	nc.QueueSubscribe("tasks.process", "worker-group", func(msg *nats.Msg) {
		fmt.Printf("Worker处理任务: %s\n", string(msg.Data))
		msg.Ack() // JetStream模式需要ACK
	})

	// 5. JetStream模式（持久化流处理）
	js, err := nc.JetStream()
	if err != nil {
		log.Fatal(err)
	}

	// 创建Stream
	streamName := "ORDERS"
	_, err = js.AddStream(&nats.StreamConfig{
		Name:     streamName,
		Subjects: []string{"orders.>"},
		MaxAge:   24 * time.Hour, // 保留24小时
	})
	if err != nil {
		log.Printf("Stream可能已存在: %v", err)
	}

	// 发布到Stream
	_, err = js.Publish("orders.new", []byte("order data"))
	if err != nil {
		log.Fatal(err)
	}

	// 订阅Stream（Push模式）
	js.Subscribe("orders.>", func(msg *nats.Msg) {
		fmt.Printf("Stream消息: %s\n", string(msg.Data))
		msg.Ack()
	}, nats.Durable("order-processor"))

	// 保持运行
	time.Sleep(10 * time.Second)
}

func processEvent(msg *nats.Msg) {
	// 事件处理逻辑
	// 注意：NATS Core模式是fire-and-forget，无ACK机制
	// JetStream模式需要手动ACK
}

/**
 * 最佳实践：
 * 1. 使用多个服务器地址实现自动故障转移
 * 2. 实现重连处理，利用自动重连机制
 * 3. 使用队列组实现负载均衡
 * 4. JetStream模式用于需要持久化的场景
 *
 * 参考文档：
 * - [01-05-程序设计模式分析](../01-基础概念与对比分析/01-05-程序设计模式分析.md)
 * - [02-03-程序设计模式场景化应用](../02-场景驱动架构设计/02-03-程序设计模式场景化应用.md)
 * - [04-03-负载动态响应与故障传播](../04-动态系统论与演化分析/04-03-负载动态响应与故障传播.md)
 */
